absl/absl.h

```diff
@@ -5,10 +5,41 @@
 
 namespace nda {
 
-// interval -> string.
+// interval -> string as the half-open interval [begin, end).
+//
+// Stringifies only the values, not whether they are static or dynamic.
 template <typename Sink, index_t Min_ = dynamic, index_t Extent_ = dynamic>
-void AbslStringify(Sink& sink, const nda::interval<Min_, Extent_>& p) {
-  absl::Format(&sink, "[%v, %v)", p.min(), p.min() + p.extent());
+void AbslStringify(Sink& sink, const interval<Min_, Extent_>& i) {
+  absl::Format(&sink, "[%v, %v)", i.min(), i.min() + i.extent());
+}
+
+// dim -> string as "dim(min, extent, stride)".
+//
+// Stringifies only the values, not whether they are static or dynamic.
+template <typename Sink, index_t Min_ = dynamic, index_t Extent_ = dynamic,
+    index_t Stride_ = dynamic>
+void AbslStringify(Sink& sink, const dim<Min_, Extent_, Stride_>& d) {
+  absl::Format(&sink, "dim(%v, %v, %v)", d.min(), d.extent(), d.stride());
+}
+
+// shape -> string as "shape<`rank`>(dims...).
+//
+// Stringifies only the rank and each dim's values, not whether they are static
+// or dynamic.
+//
+// TODO(jiawen): This can be implemented at compile-time using index_sequence
+// (and a fold expression in C++17) but is not worth the reduced readability.
+template <typename Sink, class... Dims>
+void AbslStringify(Sink& sink, const shape<Dims...>& sh) {
+  absl::Format(&sink, "shape<%d>(", sh.rank());
+  for (int i = 0; i < sh.rank(); ++i) {
```

Comment:
`absl::StrJoin` says it accepts tuples. So I think this can literally just be:
```
absl::Format(&sink, "%v", absl::StrJoin(sh.dims(), ",");
```
Maybe there's some more concise way to get a string into `sink` too, not sure.

Reply:
TIL `StrJoin` works on tuples, cool.

Re: more concise way of getting a string into `sink` - is `sink.Append(")")` below too verbose? It seems to be what the absl tests use.

Reply:
No the append is fine, I wasn't thinking when I read that.

Reply:
Done.